File: basic_dsa_two_hundred_eighty_seven.java
/**
 * Basic DSA Problems - Product of Array Except Self
 * Return array where output[i] = product of all elements except nums[i]
 */
public class basic_dsa_two_hundred_eighty_seven {

    // Problem: output[i] = product of all nums except nums[i]. O(n) time, O(1) extra space.
    // Cannot use division.

    /**
     * Two passes: left products then right. Or prefix * suffix in one output array.
     * Time O(n), Space O(1) excluding output
     */
    public static int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] out = new int[n];
        out[0] = 1;
        for (int i = 1; i < n; i++) out[i] = out[i - 1] * nums[i - 1];
        int right = 1;
        for (int i = n - 1; i >= 0; i--) {
            out[i] *= right;
            right *= nums[i];
        }
        return out;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Product of Array Except Self");
        System.out.println("=".repeat(60));

        System.out.println("[1,2,3,4] -> " + java.util.Arrays.toString(productExceptSelf(new int[]{1, 2, 3, 4})));
        System.out.println("[-1,1,0,-3,3] -> " + java.util.Arrays.toString(productExceptSelf(new int[]{-1, 1, 0, -3, 3})));
    }
}

File: basic_dsa_two_hundred_eighty_seven.cpp
/**
 * Basic DSA - Convert Sorted Array to BST (LeetCode 108)
 * Build height-balanced BST from sorted array.
 */
#include <iostream>
#include <vector>

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* build(std::vector<int>& nums, int lo, int hi) {
    if (lo > hi) return nullptr;
    int mid = lo + (hi - lo) / 2;
    TreeNode* root = new TreeNode(nums[mid]);
    root->left = build(nums, lo, mid - 1);
    root->right = build(nums, mid + 1, hi);
    return root;
}

TreeNode* sortedArrayToBST(std::vector<int>& nums) {
    return build(nums, 0, nums.size() - 1);
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Convert Sorted Array to BST\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> v1 = {-10, -3, 0, 5, 9};
    TreeNode* r1 = sortedArrayToBST(v1);
    std::cout << "[-10,-3,0,5,9] -> root=" << r1->val
              << " L=" << r1->left->val << " R=" << r1->right->val << "\n";

    std::vector<int> v2 = {1, 3};
    TreeNode* r2 = sortedArrayToBST(v2);
    std::cout << "[1,3] -> root=" << r2->val << "\n";
}
